process/rating/ratingsData.go

```diff
@@ -18,70 +21,177 @@ type RatingsData struct {
 			process.ErrConsecutiveMissedBlocksPenaltyLowerThanOne,
 			settings.ShardChain.ConsecutiveMissedBlocksPenalty)
 	}
-	if settings.ShardChain.ProposerDecreaseRatingStep > 0 || settings.ShardChain.ValidatorDecreaseRatingStep > 0 {
-		return nil, fmt.Errorf("%w: shardChain decrease steps - proposer: %v, validator: %v",
-			process.ErrDecreaseRatingsStepPositive,
-			settings.ShardChain.ProposerDecreaseRatingStep,
-			settings.ShardChain.ValidatorDecreaseRatingStep)
+	if settings.ShardChain.ProposerDecreaseFactor > -1 || settings.ShardChain.ValidatorDecreaseFactor > -1 {
+		return fmt.Errorf("%w: shardChain decrease steps - proposer: %v, validator: %v",
+			process.ErrDecreaseRatingsStepMoreThanMinusOne,
+			settings.ShardChain.ProposerDecreaseFactor,
+			settings.ShardChain.ValidatorDecreaseFactor)
 	}
-
-	if settings.MetaChain.ProposerDecreaseRatingStep > 0 || settings.MetaChain.ValidatorDecreaseRatingStep > 0 {
-		return nil, fmt.Errorf("%w: metachain decrease steps - proposer: %v, validator: %v",
-			process.ErrDecreaseRatingsStepPositive,
-			settings.MetaChain.ProposerDecreaseRatingStep,
-			settings.MetaChain.ValidatorDecreaseRatingStep)
+	if settings.MetaChain.ProposerDecreaseFactor > -1 || settings.MetaChain.ValidatorDecreaseFactor > -1 {
+		return fmt.Errorf("%w: metachain decrease steps - proposer: %v, validator: %v",
+			process.ErrDecreaseRatingsStepMoreThanMinusOne,
+			settings.MetaChain.ProposerDecreaseFactor,
+			settings.MetaChain.ValidatorDecreaseFactor)
 	}
+	return nil
+}
 
-	chances := make([]process.SelectionChance, 0)
-	for _, chance := range settings.General.SelectionChances {
-		chances = append(chances, &SelectionChance{
-			MaxThreshold:  chance.MaxThreshold,
-			ChancePercent: chance.ChancePercent,
-		})
+func computeRatingStep(
+	shardSize uint32,
+	consensusSize uint32,
+	roundTimeMilis uint64,
+	startRating uint32,
+	maxRating uint32,
+	hoursToMaxRatingFromStartRating uint32,
+	proposerDecreaseFactor float32,
+	validatorDecreaseFactor float32,
+	consecutiveMissedBlocksPenalty float32,
+	proposerValidatorImportance float32,
+) (process.RatingsStepHandler, error) {
+	blocksProducedInHours := uint64(hoursToMaxRatingFromStartRating*milisecondsInHour) / roundTimeMilis
+	ratingDifference := maxRating - startRating
+
+	proposerProbability := float32(blocksProducedInHours) / float32(shardSize)
+	validatorProbability := proposerProbability * float32(consensusSize)
+
+	totalImportance := proposerValidatorImportance + 1
+
+	ratingFromProposer := float32(ratingDifference) * (proposerValidatorImportance / totalImportance)
+	ratingFromValidator := float32(ratingDifference) * (1 / totalImportance)
+
+	proposerIncrease := ratingFromProposer / proposerProbability
+	validatorIncrease := ratingFromValidator / validatorProbability
+	proposerDecrease := proposerIncrease * proposerDecreaseFactor
+	validatorDecrease := validatorIncrease * validatorDecreaseFactor
+
+	if proposerIncrease > math.MaxInt32 {
```

Comment:
this could never happen. Over and UnderFlow protection has to be done before the operation. You could do all the computation with big.Int as it is only called in constructor.

Reply:
proposerIncrease is a float32 and it can happen. I wrote some unit Tests

Reply:
I think float32 goes somewhere up to 2^127, although not sure about the precision, so the check should be ok.